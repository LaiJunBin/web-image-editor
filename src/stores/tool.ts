import { defineStore } from 'pinia'
import { Tool } from '@/models/Tool'
import { ref, type Ref } from 'vue'

export const useToolStore = defineStore('tool', () => {
  const tool = ref<Tool>() as Ref<Tool>
  const setTool = (newTool: Tool) => {
    tool.value = newTool
  }

  return { tool, setTool }
})
