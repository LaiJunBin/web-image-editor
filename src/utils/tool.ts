interface DrawOptions {
  startX: number
  startY: number
  endX: number
  endY: number
  color: string
  lineWidth: number
}

export function draw(ctx: CanvasRenderingContext2D, options: DrawOptions) {
  const { startX, startY, endX, endY, color, lineWidth } = options

  ctx.beginPath()
  ctx.strokeStyle = color
  ctx.moveTo(startX, startY)
  ctx.lineTo(endX, endY)
  ctx.lineWidth = lineWidth
  ctx.lineCap = 'round'
  ctx.stroke()
  ctx.closePath()
}
