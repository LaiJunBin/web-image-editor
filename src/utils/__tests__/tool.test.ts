import { draw } from '../tool'

const canvas = document.createElement('canvas')
const ctx = canvas.getContext('2d')!
vi.mock('@/stores/layer', async () => {
  return {
    useLayerStore: () => ({
      recordLayer: {
        ctx,
        save: vi.fn(),
        restore: vi.fn()
      }
    })
  }
})

describe('tool utils', () => {
  beforeEach(() => {
    ctx.__clearEvents()
    ctx.__clearDrawCalls()
  })

  test('draw', () => {
    draw(ctx, {
      startX: 0,
      startY: 0,
      endX: 100,
      endY: 100,
      color: '#333',
      lineWidth: 10
    })

    expect(ctx.__getDrawCalls()).toMatchSnapshot()
  })
})
