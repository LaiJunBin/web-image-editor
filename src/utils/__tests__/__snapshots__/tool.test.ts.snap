// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tool utils > draw 1`] = `
[
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        {
          "props": {
            "x": 0,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 100,
            "y": 100,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
